src/server/master_server/metadata_manager.cc

```diff
@@ -0,0 +1,132 @@
+  // The reason that we acquire the global lock is that we need to synchronization
+  // between write and read from the fileMetadata collection.
+  if(fileMetadata_.contains(pathname)) {
+    return google::protobuf::util::Status(
+             google::protobuf::util::error::OUT_OF_RANGE,
+             "File metadata already exists for " + pathname);
+  }
+   
+  fileMetadata_[pathname] = std::make_shared<FileMetadata>();
+  // Initialize the filename
+  fileMetadata_[pathname]->set_filename(pathname);
+  return google::protobuf::util::Status::OK;
+}
+
+bool MetadataManager::ExistFileMetadata(const std::string& pathname) const {
+  absl::ReaderMutexLock anchor(fileMetadataLock_);
+  return fileMetadata_.contains(pathname);
+}
+
+google::protobuf::util::StatusOr<std::shared_ptr<FileMetadata>> 
+  MetadataManager::GetFileMetadata(
+  
+  const std::string& pathname) const {
+  // readLock the global lock and retrieve the filemetadata. The reason for a
+  // readLock is because we are not mutating anything in the fileMetadata_.
+  absl::ReaderMutexLock anchor(fileMetadataLock_);
+  if (!fileMetadata_.contains(pathname)) {
+     return google::protobuf::util::Status(
+              google::protobuf::util::error::OUT_OF_RANGE,
+              "File metadata does not exist: " + pathname);
+  }
+  return fileMetadata_.at(pathname);
+}
+
+google::protobuf::util::StatusOr<std::string> 
+  MetadataManager::CreateChunkHandle(const std::string& pathname, 
+                                     uint32_t chunk_index) {
+  // Step 1. readlock the parent directories
+  ParentLocksAnchor parentLockAnchor(lockManager_, pathname);
+  if (!parentLockAnchor.ok()) {
+    // If this operation fails, which means some of the parent directory does not
+    // exist, we return false
+    return parentLockAnchor.status();
+  }
+
+  // Step 2. readlock the global lock, fetch the data and unlock readerlock
+  auto fdataRes(GetFileMetadata(pathname));
+  if (!fdataRes.ok()) {
+    return fdataRes.status();  
+  }
+  auto fdata(fdataRes.ValueOrDie());
+
+  // Step 3. writelock the lock for this path
+  auto fLockRes(lockManager_->FetchLock(pathname)); 
+  if (!fLockRes.ok()) {
+    return fLockRes.status();
+  }
+ 
+  auto fLock(fLockRes.ValueOrDie());
+  absl::WriterMutexLock anchor(fLock);
+
+  // Step 4. compute a new chunk handle, and insert the (chunk_index, chunkHandle)
+  //         pair to fileMetadata_
+  std::string newChunkHandle(AllocateNewChunkHandle());
+  fdata->set_filename(pathname);
+  auto& chunk_handle_map(*fdata->mutable_chunk_handles());
+  
+  // Return null UIDD if this chunk_index exists
+  if(chunk_handle_map.contains(chunk_index)) {
+     return google::protobuf::util::Status(
+              google::protobuf::util::error::OUT_OF_RANGE,
+                 "Chunk " + std::to_string(chunk_index) + "already exists in file " 
+                 + pathname);
+  }
+   
+  chunk_handle_map[chunk_index]= newChunkHandle;
```

Comment:
```
 chunk_handle_map[chunk_index]= AllocateNewChunkHandle();
```

Reply:
Again not storing the AllocateNewChunkHandle() in a string means we have to perform the lookup twice, or we'd have to do something like `return chunk_handle_map[chunk_index]= AllocateNewChunkHandle();` ?